File: app/src/main/java/org/dhis2/utils/custom_views/OptionSetDialog.java
package org.dhis2.utils.custom_views;

import android.app.Dialog;
import android.content.DialogInterface;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.view.Window;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.databinding.DataBindingUtil;
import androidx.fragment.app.DialogFragment;
import androidx.recyclerview.widget.RecyclerView;

import com.jakewharton.rxbinding2.widget.RxTextView;

import org.dhis2.R;
import org.dhis2.data.forms.dataentry.fields.spinner.SpinnerViewModel;
import org.dhis2.data.tuples.Trio;
import org.dhis2.databinding.DialogOptionSetBinding;
import org.dhis2.utils.EndlessRecyclerViewScrollListener;
import org.hisp.dhis.android.core.option.OptionModel;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.processors.FlowableProcessor;
import io.reactivex.schedulers.Schedulers;
import timber.log.Timber;

public class OptionSetDialog extends DialogFragment {

    private static OptionSetDialog instace;
    private DialogOptionSetBinding binding;
    private CompositeDisposable disposable;
    //1st param is text to search, 2nd param is uid of optionSet,3rd param is page
    private FlowableProcessor<Trio<String, String, Integer>> processor;
    private OptionSetAdapter adapter;
    private SpinnerViewModel optionSet;

    private OptionSetOnClickListener listener;
    private View.OnClickListener cancelListener;
    private View.OnClickListener clearListener;

    private EndlessRecyclerViewScrollListener endlessScrollListener;

    public static OptionSetDialog newInstance() {
        if (instace == null) {
            instace = new OptionSetDialog();
        }
        return instace;
    }

    public static Boolean isCreated() {
        return instace != null;
    }

    @NotNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        Dialog dialog = super.onCreateDialog(savedInstanceState);
        dialog.getWindow().requestFeature(Window.FEATURE_NO_TITLE);
        dialog.getWindow().setBackgroundDrawableResource(android.R.color.transparent);

        return dialog;
    }

    @Override
    public void onCancel(@NonNull DialogInterface dialog) {
        instace = null;
        disposable.clear();
        super.onCancel(dialog);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, Bundle savedInstanceState) {
        binding = DataBindingUtil.inflate(inflater, R.layout.dialog_option_set, container, false);


        binding.title.setText(optionSet.label());
        disposable = new CompositeDisposable();

        endlessScrollListener = new EndlessRecyclerViewScrollListener(binding.recycler.getLayoutManager(), 2, 0) {
            @Override
            public void onLoadMore(int page, int totalItemsCount, RecyclerView view) {
                processor.onNext(Trio.Companion.create(binding.txtSearch.getText().toString(), optionSet.optionSet(), page));
            }
        };
        binding.recycler.addOnScrollListener(endlessScrollListener);
        adapter = new OptionSetAdapter(listener);
        adapter.setOptions(new ArrayList<>(), endlessScrollListener.getCurrentPage());
        binding.recycler.setAdapter(adapter);

        processor.onNext(Trio.Companion.create("", optionSet.optionSet(), 0));

        binding.clearButton.setOnClickListener(clearListener);
        binding.cancelButton.setOnClickListener(cancelListener);
        disposable.add(RxTextView.textChanges(binding.txtSearch)
                .debounce(500, TimeUnit.MILLISECONDS)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(data -> {
                            endlessScrollListener.resetState();
                            processor.onNext(Trio.Companion.create(data.toString(), optionSet.optionSet(), 0));
                        },
                        Timber::e
                ));

        return binding.getRoot();

    }

    @Override
    public void dismiss() {
        instace = null;
        disposable.clear();
        super.dismiss();
    }

    public OptionSetDialog setOnClick(OptionSetOnClickListener listener) {
        this.listener = listener;
        return this;
    }

    public OptionSetDialog setOptions(List<OptionModel> options) {
        if (options.isEmpty() && adapter.isLastItemLoading(adapter.getItemCount())) {
            adapter.remove();
        } else {
            adapter.setOptions(options, endlessScrollListener.getCurrentPage());
        }
        return this;
    }

    public OptionSetDialog setOptionSetUid(SpinnerViewModel view) {
        this.optionSet = view;
        return this;
    }

    public OptionSetDialog setProcessor(FlowableProcessor<Trio<String, String, Integer>> processor) {
        this.processor = processor;
        return this;
    }

    public OptionSetDialog setCancelListener(View.OnClickListener cancelListener) {
        this.cancelListener = cancelListener;
        return this;
    }

    public OptionSetDialog setClearListener(View.OnClickListener clearListener) {
        this.clearListener = clearListener;
        return this;
    }
}

File: app/src/main/java/org/dhis2/utils/optionset/OptionSetDialog.kt
package org.dhis2.utils.optionset

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.view.Window
import androidx.databinding.DataBindingUtil
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.FragmentManager
import androidx.lifecycle.LiveData
import androidx.lifecycle.Observer
import androidx.paging.PagedList
import com.jakewharton.rxbinding2.widget.RxTextView
import io.reactivex.Observable
import javax.inject.Inject
import org.dhis2.App
import org.dhis2.R
import org.dhis2.data.forms.dataentry.fields.spinner.SpinnerViewModel
import org.dhis2.data.forms.dataentry.tablefields.spinner.SpinnerViewModel as TableSpinnerViewModel
import org.dhis2.databinding.DialogOptionSetBinding
import org.dhis2.utils.Constants
import org.dhis2.utils.customviews.OptionSetOnClickListener
import org.hisp.dhis.android.core.option.Option

class OptionSetDialog : DialogFragment(), OptionSetView {

    private lateinit var binding: DialogOptionSetBinding

    @Inject
    lateinit var presenter: OptionSetPresenter

    private var adapter: OptionSetAdapter? = null

    var optionSet: SpinnerViewModel? = null
    var optionSetTable: TableSpinnerViewModel? = null
    var listener: OptionSetOnClickListener? = null
    var clearListener: View.OnClickListener? = null
    var defaultSize: Int = 0

    var isDialogShown = false
        private set

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val dialog = super.onCreateDialog(savedInstanceState)
        dialog.window!!.requestFeature(Window.FEATURE_NO_TITLE)
        dialog.window!!.setBackgroundDrawableResource(android.R.color.transparent)

        return dialog
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = DataBindingUtil.inflate(inflater, R.layout.dialog_option_set, container, false)

        binding.title.text = if (optionSet != null) optionSet?.label() else optionSetTable?.label()
        binding.title.text = if (optionSet != null) optionSet?.label() else optionSetTable?.label()

        if (optionSet != null) {
            presenter.init(optionSet!!)
        } else if (optionSetTable != null) {
            presenter.init(optionSetTable!!)
        }

        adapter = OptionSetAdapter(
            OptionSetOnClickListener {
                listener?.onSelectOption(it)
                this.dismiss()
            }
        )

        binding.recycler.adapter = adapter

        binding.clearButton.setOnClickListener { view ->
            clearListener?.onClick(view)
            this.dismiss()
        }
        binding.cancelButton.setOnClickListener { this.dismiss() }

        return binding.root
    }

    override fun onCancel(dialog: DialogInterface) {
        presenter.onDetach()
        super.onCancel(dialog)
    }

    override fun setLiveData(data: LiveData<PagedList<Option>>?) {
        data?.observe(
            this,
            Observer {
                adapter?.submitList(it)
            }
        )
    }

    fun create(context: Context) {
        (context.applicationContext as App)
            .userComponent()!!.plus(OptionSetModule(this)).inject(this)
        defaultSize = context.getSharedPreferences(Constants.SHARE_PREFS, Context.MODE_PRIVATE)
            .getInt(Constants.OPTION_SET_DIALOG_THRESHOLD, 15)
    }

    override fun show(manager: FragmentManager, tag: String?) {
        isDialogShown = true
        super.show(manager, tag)
    }

    override fun showDialog(): Boolean {
        return presenter.getCount(
            if (optionSet != null) {
                optionSet!!.optionSet()
            } else {
                optionSetTable!!.optionSet()
            }
        )!! > defaultSize
    }

    override fun dismiss() {
        presenter.onDetach()
        if (isDialogShown) {
            isDialogShown = false
            super.dismiss()
        }
    }

    override fun searchSource(): Observable<CharSequence> {
        return RxTextView.textChanges(binding.txtSearch)
            .startWith("")
    }

    companion object {
        val TAG: String = OptionSetDialog::class.java.name
    }
}
